src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets

```diff
@@ -115,27 +116,33 @@
           DependsOnTargets="Protobuf_Compile"
           Condition=" '$(Language)' == 'C#' " />
 
-  <Target Name="_Protobuf_SelectFiles">
+  <Target Name="_Protobuf_SetProtoRoot">
     <!-- Guess .proto root for the files. Whenever the root is set for a file explicitly,
          leave it as is. Otherwise, for files under the project directory, set the root
          to "." for the project's directory, as it is the current when compiling; for the
-         files outside of project directory, use each .proto file's directory as the root. -->
+         files outside of project directory, use each .proto file's directory as the root 
+		 or Protobuf_ProtoRoot if set. -->
     <FindUnderPath Path="$(MSBuildProjectDirectory)"
                    Files="@(Protobuf->WithMetadataValue('ProtoRoot',''))">
       <Output TaskParameter="InPath" ItemName="_Protobuf_NoRootInProject"/>
       <Output TaskParameter="OutOfPath" ItemName="_Protobuf_NoRootElsewhere"/>
     </FindUnderPath>
+	
+    <ItemGroup>
+      <Protobuf>
+		<!-- In-project files will have ProtoRoot='.'. -->
+        <ProtoRoot Condition=" '%(ProtoRoot)' == '' and '@(_Protobuf_NoRootInProject)' != '' ">.</ProtoRoot>
```

Comment:
@jtattermusch In MSBuild, you can modify or update the metadata of the ItemGroup with an entry under the ItemGroup node. 

So in this case, it seems like the logic wants to set the `ProtoRoot` metadata of `Protobuf` items if it's not already set based on `_Protobuf_NoRootInProject` and `_Protobuf_NoRootElsewhere `. However I'm not sure if it would work correctly.  `_Protobuf_NoRootInProject` and `_Protobuf_NoRootElsewhere ` are ItemGroups so '@(_Protobuf_NoRootElsewhere)' != '' would evaluate to true as long as those items contain a single element. However, the logic doesn't take into account whether the `Protobuf`item belongs to one group or the other.

For example, currently, if a.proto was placed in `_Protobuf_NoRootInProject` and b.proto was placed in  `_Protobuf_NoRootElsewhere` both would have their ProtoRoot set to `%(RelativeDir)` since both conditions are true and the second metadata update overrides the first.

Reply:
Are you sure? I‘m not too deep into MSBuild, but based on what I saw in the use cases, it seems that @(x) in a condition checks if %(Identity) is part of the ItemGroup and returns the item. I will try to verify the behavior. Thanks for the input!

Reply:
I made a test with the following input:
```
  <ItemGroup>
    <Protobuf Include="protos\config.proto" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\field_identifier.proto" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\observer\notification.proto" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\predicate.proto" ProtoRoot="$(LIProtoPath)" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\predicate_request.proto" ProtoRoot="$(LIProtoPath)" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\value.proto" ProtoRoot="$(LIProtoPath)" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\observer\register.proto" ProtoRoot="$(LIProtoPath)" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\observer\unregister.proto" ProtoRoot="$(LIProtoPath)" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\observer\watch.proto" ProtoRoot="$(LIProtoPath)" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\observer\v1\observer.proto" ProtoRoot="$(LIProtoPath)" />
    <Protobuf Include="$(LIProtoPath)\jungheinrich\descriptor.proto" Access="Internal" />
    <Protobuf Include="$(LIProtoPath)\grpc\lb\v1\load_balancer.proto" Access="Internal" />
  </ItemGroup>
```

The output of `<Warning Text="%(Protobuf.Identity): %(Protobuf.ProtoRoot)" />` after applying the logic was:
``` 
_Protobuf_NoRootInProject: protos\config.proto
_Protobuf_NoRootElsewhere: $(LIProtoPath)\jungheinrich\field_identifier.proto;$(LIProtoPath)\jungheinrich\observer\notification.proto;$(LIProtoPath)\jungheinrich\descriptor.proto;$(LIProtoPath)\grpc\lb\v1\load_balancer.proto

protos\config.proto: .
$(LIProtoPath)\jungheinrich\field_identifier.proto: $(LIProtoPath)\jungheinrich\
$(LIProtoPath)\jungheinrich\observer\notification.proto: $(LIProtoPath)\jungheinrich\observer\
$(LIProtoPath)\jungheinrich\predicate.proto: $(LIProtoPath)
$(LIProtoPath)\jungheinrich\predicate_request.proto: $(LIProtoPath)
$(LIProtoPath)\jungheinrich\value.proto: $(LIProtoPath)
$(LIProtoPath)\jungheinrich\observer\register.proto: $(LIProtoPath)
$(LIProtoPath)\jungheinrich\observer\unregister.proto: $(LIProtoPath)
$(LIProtoPath)\jungheinrich\observer\watch.proto: $(LIProtoPath)
$(LIProtoPath)\jungheinrich\observer\v1\observer.proto: $(LIProtoPath)
$(LIProtoPath)\grpc\lb\v1\load_balancer.proto: $(LIProtoPath)\grpc\lb\v1\
```

So `@(_Protobuf_NoRootElsewhere)` in the Condition of `ProtoRoot` checks if `%(Identity)` is contained in the item group.

Reply:
Hmm it seems like collecting binlogs of before and after the change is needed to verify what the logic is doing. As a test, I've ran the following logic:
```xml<Project>
    <Target Name="Test">
        <ItemGroup>
            <Protobuf Include="a.proto" Group="_Protobuf_NoRootInProject" ProtoRoot="" />
            <Protobuf Include="b.proto" Group="_Protobuf_NoRootElsewhere" ProtoRoot="" />
            <Protobuf Include="c.proto" Group="" ProtoRoot="Preset"  />

            <_Protobuf_NoRootInProject Include="@(Protobuf)" Condition="'%(Group)' == '_Protobuf_NoRootInProject'" />
            <_Protobuf_NoRootElsewhere Include="@(Protobuf)" Condition="'%(Group)' == '_Protobuf_NoRootElsewhere'" />

            <Protobuf>
                <ProtoRoot Condition="'%(ProtoRoot)' == '' and '@(_Protobuf_NoRootInProject)' != '' ">.</ProtoRoot>
                <ProtoRoot Condition="'%(ProtoRoot)' == '' and '@(_Protobuf_NoRootElsewhere)' != '' ">RelativeDir</ProtoRoot>
            </Protobuf>
        </ItemGroup>

        <Message Importance="High" Text="_Protobuf_NoRootInProject: @(_Protobuf_NoRootInProject)" />
        <Message Importance="High" Text="_Protobuf_NoRootElsewhere: @(_Protobuf_NoRootElsewhere)" />

        <Message Importance="High" Text="Protobuf: %(Protobuf.Identity) - ProtoRoot: %(Protobuf.ProtoRoot)" />
    </Target>
</Project>
```

The output is 
```
  _Protobuf_NoRootInProject: a.proto
  _Protobuf_NoRootElsewhere: b.proto
  Protobuf: a.proto - ProtoRoot: RelativeDir
  Protobuf: b.proto - ProtoRoot: RelativeDir
  Protobuf: c.proto - ProtoRoot: Preset
```

Which highlights my concern with the current logic.

Reply:
I will check why in my test `protos\config.proto: .` was working and in yours not.

Reply:
@JunTaoLuo I found out why my test was working. You were right, it is overwritten. But for `protos\config.proto` the `%(RelativeDir)` was `.`.

I will update the target file to use the original logic but with a separate item group. I only wanted to avoid having a duplicate of the group until necessary.